tests/unit/iteratorFilter.cpp

```diff
@@ -0,0 +1,385 @@
+
+/*
+ *   Copyright 2021 Huawei Technologies Co., Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <vector>
+#include <numeric>
+#include <iostream>
+
+#include <graphblas.hpp>
+#include <graphblas/utils/iterators/filter.hpp>
+
+
+template< typename FwdIt >
+void testOneOut(
+	const size_t &n, const size_t filtered, const size_t expected,
+	FwdIt &it, const FwdIt &end,
+	grb::RC &rc
+) {
+	rc = grb::SUCCESS;
+	size_t count = 0;
+	for( ; it != end; ++it, ++count ) {
+		if( *it != *(it.operator->()) ) {
+			std::cerr << "Error: dereference operator returns " << *it << ", "
+				<< "but dereferencing the pointer operator returns " << (*(it.operator->()))
+				<< "\n";
+			rc = grb::FAILED;
+			return;
+		}
+		if( *it == filtered ) {
+			std::cerr << "Error: found " << filtered << ", "
+				<< "which should have been filtered out\n";
+			rc = grb::FAILED;
+			return;
+		}
+	}
+	if( count != expected ) {
+		std::cerr << "Error: count should be " << n << ", "
+			<< "but the filtered iterator returned " << count << " elements instead\n";
+		rc = grb::FAILED;
+	}
+}
```

Comment:
```suggestion
template< typename FwdIt >
void testOneOut(
	const size_t &n, const size_t filtered,
	FwdIt &it, const FwdIt &end,
	grb::RC &rc
) {
	const size_t expected_size = n - std::count( it, end, filtered );
	
	// Technically, n could also be replaced by:
	// const size_t n = std::distance( it, end );
	
	rc = grb::SUCCESS;
	size_t count = 0;
	for( ; it != end; ++it, ++count ) {
		if( *it != *(it.operator->()) ) {
			std::cerr << "Error: dereference operator returns " << *it << ", "
				<< "but dereferencing the pointer operator returns " << (*(it.operator->()))
				<< "\n";
			rc = grb::FAILED;
			return;
		}
		if( *it == filtered ) {
			std::cerr << "Error: found " << filtered << ", "
				<< "which should have been filtered out\n";
			rc = grb::FAILED;
			return;
		}
	}
	if( count != expected_size ) {
		std::cerr << "Error: count should be " << expected_size << ", "
			<< "but the filtered iterator returned " << count << " elements instead\n";
		rc = grb::FAILED;
	}
}
```

Reply:
I'm not sure what the difference is between the blocks, OK to ignore?

Reply:
The *expected* is computed in the function instead of begin passed as an argument

Reply:
Ah I see, sneaky hahah. I prefer the old variant because it explicitly forces the unit tester to think about what they expect=)

Reply:
Sounds good